channeldb/kvdb/etcd/stm.go

```diff
@@ -0,0 +1,736 @@
+// options then creates the stm object and resets it before returning.
+func makeSTM(cli *v3.Client, manual bool, so ...STMOptionFunc) *stm {
+	opts := &STMOptions{
+		ctx: cli.Ctx(),
+	}
+
+	// Apply all functional options.
+	for _, fo := range so {
+		fo(opts)
+	}
+
+	s := &stm{
+		client:   cli,
+		manual:   manual,
+		options:  opts,
+		prefetch: make(map[string]stmGet),
+	}
+
+	// Reset read and write set.
+	s.Rollback()
+
+	return s
+}
+
+// runSTM implements the run loop of the STM, running the
+// apply func, catching errors and handling commit. The loop
+// will quit on every error except CommitError which is used
+// to indicate a necessary retry.
+func runSTM(s *stm, apply func(STM) error) error {
+	out := make(chan error, 1)
+
+	go func() {
+		defer func() {
+			// Recover DatabaseError panics so
+			// we can return them.
+			if r := recover(); r != nil {
+				e, ok := r.(DatabaseError)
+				if !ok {
+					// Unknown panic.
+					panic(r)
+				}
+
+				// Return the error.
+				out <- e.Unwrap()
+			}
+		}()
+
+		var err error
+
+		// In a loop try to apply and commit and roll back
+		// if the database has changed (CommitError).
+		for {
+			// Abort STM if there was an application error.
+			if err = apply(s); err != nil {
+				break
+			}
+
+			err = s.Commit()
+
+			// Re-apply only upon commit error
+			// (meaning the database was changed).
+			if _, ok := err.(CommitError); !ok {
+				// Anything that's not a CommitError
+				// aborts the STM run loop.
+				break
+			}
+
+			// Rollback before trying to re-apply.
+			s.Rollback()
+		}
+
+		// Return the error to the caller.
+		out <- err
+	}()
+
+	return <-out
+}
+
+// add inserts a txn response to the read set. This is
+// useful when the txn fails due to conflict where the
+// txn response can be used to prefetch key/values.
+func (rs readSet) add(txnResp *v3.TxnResponse) {
+	for _, resp := range txnResp.Responses {
+		getResp := (*v3.GetResponse)(resp.GetResponseRange())
+		for _, kv := range getResp.Kvs {
+			rs[string(kv.Key)] = stmGet{
+				val: string(kv.Value),
+				rev: kv.ModRevision,
+			}
+		}
+	}
+}
+
+// gets is a helper to create an op slice for transaction
+// construction.
+func (rs readSet) gets() []v3.Op {
+	ops := make([]v3.Op, 0, len(rs))
+
+	for k := range rs {
+		ops = append(ops, v3.OpGet(k))
+	}
+
+	return ops
+}
+
+// cmps returns a cmp list testing values in read set didn't change
+func (rs readSet) cmps() []v3.Cmp {
+	cmps := make([]v3.Cmp, 0, len(rs))
+	for key, getValue := range rs {
+		cmps = append(cmps, v3.Compare(v3.ModRevision(key), "=", getValue.rev))
+	}
+
+	return cmps
+}
+
+// cmps returns a cmp list testing no writes have happened past rev
+func (ws writeSet) cmps(rev int64) []v3.Cmp {
+	cmps := make([]v3.Cmp, 0, len(ws))
+	for key := range ws {
+		cmps = append(cmps, v3.Compare(v3.ModRevision(key), "<", rev))
+	}
+
+	return cmps
+}
+
+// puts is the list of ops for all pending writes
```

Comment:
Style nit: missing a period here and the above two methods.

Reply:
done